README
This package contains example programs for the FreeType 2 library.

You can build it either using the Meson build tool (see `README.meson`
for  more details),  or  using GNU  Make,  following the  instructions
below.

WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING

  The Makefile contained in this  directory assume that the FreeType 2
  library sources are located in `../freetype`.  If you downloaded one
  of the stable  FreeType 2 source packages from our  server, you most
  probably have to rename its directory, for example:

    mv freetype-2.11.0 freetype       on Unix
    rename freetype-2.11.0 freetype   on Windows

WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING


  First compile the FreeType 2 library, then say `make`.  However, you
  must rename  the directory of  the FreeType 2 library  to `freetype`
  (or create a symlink) before doing this.

  Note  that  the  demonstration  programs  include  a  tiny  graphics
  sub-system that includes 'drivers' to display Windows on Win32, X11,
  BeOS, Mac, and  OS/2.  The build system  should automatically detect
  which driver to use based on the current platform.

  UNIX USERS
  ==========

    X11 issues
    ----------

    When building the demos, the build system tries to detect your X11
    path  by  looking for  the  patterns  `X11/bin`, `X11R6/bin`,  and
    `X11R5/bin` in your current path (in  this order).  If no X11 path
    is found, some demo programs will  not be able to display graphics
    and  will fail.   If you  have X11  in an  unusual place,  use the
    `X11_PATH` make variable.  More  than one directory, if necessary,
    must be separated with spaces.  Example:

        make X11_PATH="/usr/openwin /usr/local/X11R6"

    The build system then derives  the `X11_INCLUDE` include file path
    from `X11_PATH`  by appending  `/include` to all  path components.
    It  also derives  the `X11_LIB`  library path  from `X11_PATH`  by
    appending `/lib64` and  `/lib` to all components  (in that order).
    You  might  override  those  variables similar  to  `X11_PATH`  as
    arguments to `make`.

    If you  don't have X11  at all, fix  the definition of  the `EXES`
    variable as described in the top-level Makefile.

    Recent versions of Mac OS X  no longer deliver X11 by default; you
    have  to  install 'XQuartz'  or  'xorg-server'  (the successor  of
    XQuartz).   Those  bundles  are  provided  by  distributions  like
    MacPorts or homebrew.


    Using a different build directory
    ---------------------------------

    If the `configure` script of the  FreeType 2 library is run from a
    separate build directory, it creates  a dummy `Makefile` file that
    sets some variables before including the top-level Makefile of the
    source directory.   This dummy Makefile  can also be used  for the
    'ft2demos' package to be compiled in a different build directory:

       make FT2DEMOS=1 -f /path/to/freetype/build-dir/Makefile

    If necessary, adapt  the `TOP_DIR_2` variable to make  it point to
    the 'ft2demos' source directory.

--- end of README ---


ChangeLog
2021-07-18  Werner Lemberg  <wl@gnu.org>

	* Makefile (dist): Ignore `subprojects/freetype2`.

2021-07-18  Werner Lemberg  <wl@gnu.org>

	[ftsdf] Add man page, other minor improvement.

	* man/ftsdf.1: New file.
	* src/fdsdf.c (usage): Resolution can't be changed currently.

2021-07-17  Werner Lemberg  <wl@gnu.org>

	Remove code related to warping.

	This synchronizes the demo programs with commit 2f11522a from the
	FreeType repository.

	* src/ftcommon.c (FTDemo_Hinting_Engine_Change, FTDemo_Get_Info):
	Updated.

	* src/ftdiff.c (ColumnStateRec, render_state_init,
	render_state_draw, event_help, process_event): Updated.

	* src/ftgrid.c (event_help): Updated.

	* src/ftinspect/engine/engine.cpp (Engine::Engine, Engine::update),
	src/ftinspect/engine/engine.hpp (Engine): Updated.
	* src/ftinspect/maingui.cpp (MainGUI::checkHinting,
	MainGUI::checkAutoHinting, MainGUI::createLayout),
	src/ftinspect/maingui.hpp (MainGUI): Updated.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* src/ftlint.c (main): Minor output improvement.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* meson.build: Fix version string.

2021-07-01  Werner Lemberg  <wl@gnu.org>

	* subprojects/freetype2.wrap (revision): s/head/HEAD/.

2021-06-30  Werner Lemberg  <wl@gnu.org>

	[ftlint] Improve output.

	* src/ftlint.c (main): Add header to output and remove repetitive
	strings.
	(Analyze): Updated.

2021-06-22  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Improve error reporting.

	* src/ftlint.c (main): Decorate error messages.
	(Error): Take a prefix message argumemt.

2021-06-20  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Use `ft_basename'.

2021-06-20  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Improve error handling.

	* src/ftlint.c (main): Never panic, call `Error', and proceed.
	(Error): Desipher and print an error message instead of...
	(Panic): ...removed.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Fix signed/unsigned mismatch.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Fortify custom indexing.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main, Analyze): Fix signed/unsigned mismatch.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Handle font collections.

	* src/ftlint.c (main): Increment face_index when possible.

2021-06-17  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Introduce quiet mode, remove extension guessing.

	With so many supported font formats, it does not make sense to guess
	the `ttf' or `ttc' extensions.

	* src/ftlint.c (main): Add `-q' option and improve output formatting;
	remove extension guessing.
	(Usage): Document it.

2021-06-15  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftdump.c (Print_Glyfs, Print_Programs): Handle empty glyphs.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[ftlint] Report integral X- and Y-acutance.

	Acutance is a measure of how quickly the intensity changes between
	the minimum and maximum value at the edges.  We calculate is as a
	ratio of the integrals for the absolute second derivative and the
	absolute first derivative.  It equals to 2.0 for bitmap fonts and
	approaches this value for well-hinted fonts.

	* src/ftlint.c (Analyze): Implement it.
	(main): Use it.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Report bitmap size, improve output formatting.
	(Checksum): Updated.

2021-06-14  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* src/ftlint.c (main): Implement limited indexing.
	(Usage): Document it.

2021-06-13  Sarthak Bhardwaj  <1sarthakbhardwaj@gmail.com>
            Anurag Thakur  <devanuragthakur@gmail.com>

	[ftlint] Render and calculate MD5 checksum.

	* src/ftlint.c (main): Handle the load flags and the render mode from
	the command line.
	(Checksum): Calculate MD5 sum.
	* src/md5.c, src/md5.h: MD5 source code.
	* Makefile, messon.build: Update the build systems.

2021-06-06  Anuj Verma  <anujv@iitbhilai.ac.in>

	[ftsdf] Synchronize with the change made in the `sdf` renderer.

	* src/ftsdf.c (draw): Change logic to utilize 8bit SDF instead of 16.

2021-06-04  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* graph/gblender.c (gblender_dump_stats): Fix the lookup rate.

2021-06-02  Alexei Podtelezhnikov  <apodtele@gmail.com>

	* graph/grdevice.c (grDoneSurface) [GBLENDER_STATS]: Report stats.
	* graph/gblender.c (gblender_dump_stats): Improve formatting.

2021-05-29  Werner Lemberg  <wl@gnu.org>

	Fix compilation errors and (some) warnings for clang++.

	* graph/gblblit.c (GRGB5[56]5_TO_{RED,GREEN,BLUE}): Add casts.
	(GDST_STOREP): Use `while {} (0)`.
	* graph/gblender.h (GBLENDER_LOOKUP, GBLENDER_LOOKUP_[GBR]): Use
	`while {} (0)`.
	(GBLENDER_CLOSE, GBLENDER_CHANNEL_CLOSE): Remove final semicolon.
	* graph/grblit.c (compose_pixel_full): Use `while {} (0)`.
	(extract5[56]5): Remove final semicolon.
	(blit_gray8_to_24): Fix typo.
	* graph/grfont.c (font): Make it static.

	* src/ftcommon.c (FTDemo_Icon): Don't use obfuscating comma
	operator.
	* src/ftsdf.c (main): Move definition of `flip_y` before `goto`
	instructions.

2021-05-28  Alexei Podtelezhnikov  <apodtele@gmail.com>

	[graph/win32] Fix double-dealing of some keystrokes.

	* graph/grwin32.c (gr_win32_surface_listen_event): Deprioritize
	`TranslateMessage' to avoid double-dealing of Tab, Esc, Return.

2021-05-23  David Turner  <david@freetype.org>

	* src/ftmemchk.c: Remove minor compiler warning.

	Casting a pointer to (long) isn't always correct.

2021-05-23  David Turner  <david@freetype.org>

	* src/ftpngout.c: Remove compiler warning for Mingw64 toolchain.

	The #pragma statement is only valid with Microsoft Visual C++

2021-05-23  David Turner  <david@freetype.org>

	* graph/win32/grwin32.c: Fix compiler warnings.

	- Redefine the LCS_sRGB macro to properly compile with the Mingw64
	 toolchain and the -Wmultichar warning enabled.


src/md5.c
/*
 * This is an OpenSSL-compatible implementation of the RSA Data Security, Inc.
 * MD5 Message-Digest Algorithm (RFC 1321).
 *
 * Homepage:
 * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
 *
 * Author:
 * Alexander Peslyak, better known as Solar Designer <solar at openwall.com>
 *
 * This software was written by Alexander Peslyak in 2001.  No copyright is
 * claimed, and the software is hereby placed in the public domain.
 * In case this attempt to disclaim copyright and place the software in the
 * public domain is deemed null and void, then the software is
 * Copyright (c) 2001 Alexander Peslyak and it is hereby released to the
 * general public under the following terms:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 *
 * (This is a heavily cut-down "BSD license".)
 *
 * This differs from Colin Plumb's older public domain implementation in that
 * no exactly 32-bit integer data type is required (any 32-bit or wider
 * unsigned integer data type will do), there's no compile-time endianness
 * configuration, and the function prototypes match OpenSSL's.  No code from
 * Colin Plumb's implementation has been reused; this comment merely compares
 * the properties of the two independent implementations.
 *
 * The primary goals of this implementation are portability and ease of use.
 * It is meant to be fast, but not as fast as possible.  Some known
 * optimizations are not included to reduce source code size and avoid
 * compile-time configuration.
 */

#ifndef HAVE_OPENSSL

#include <string.h>

#include <ftcommon.h>

/*
 * The basic MD5 functions.
 *
 * F and G are optimized compared to their RFC 1321 definitions for
 * architectures that lack an AND-NOT instruction, just like in Colin Plumb's
 * implementation.
 */
#define F(x, y, z)			((z) ^ ((x) & ((y) ^ (z))))
#define G(x, y, z)			((y) ^ ((z) & ((x) ^ (y))))
#define H(x, y, z)			(((x) ^ (y)) ^ (z))
#define H2(x, y, z)			((x) ^ ((y) ^ (z)))
#define I(x, y, z)			((y) ^ ((x) | ~(z)))

/*
 * The MD5 transformation for all four rounds.
 */
#define STEP(f, a, b, c, d, x, t, s) \
	(a) += f((b), (c), (d)) + (x) + (t); \
	(a) = (((a) << (s)) | (((a) & 0xffffffff) >> (32 - (s)))); \
	(a) += (b);

/*
 * SET reads 4 input bytes in little-endian byte order and stores them in a
 * properly aligned word in host byte order.
 *
 * The check for little-endian architectures that tolerate unaligned memory
 * accesses is just an optimization.  Nothing will break if it fails to detect
 * a suitable architecture.
 *
 * Unfortunately, this optimization may be a C strict aliasing rules violation
 * if the caller's data buffer has effective type that cannot be aliased by
 * MD5_u32plus.  In practice, this problem may occur if these MD5 routines are
 * inlined into a calling function, or with future and dangerously advanced
 * link-time optimizations.  For the time being, keeping these MD5 routines in
 * their own translation unit avoids the problem.
 */
#if defined(__i386__) || defined(__x86_64__) || defined(__vax__)
#define SET(n) \
	(*(MD5_u32plus *)&ptr[(n) * 4])
#define GET(n) \
	SET(n)
#else
#define SET(n) \
	(ctx->block[(n)] = \
	(MD5_u32plus)ptr[(n) * 4] | \
	((MD5_u32plus)ptr[(n) * 4 + 1] << 8) | \
	((MD5_u32plus)ptr[(n) * 4 + 2] << 16) | \
	((MD5_u32plus)ptr[(n) * 4 + 3] << 24))
#define GET(n) \
	(ctx->block[(n)])
#endif

/*
 * This processes one or more 64-byte data blocks, but does NOT update the bit
 * counters.  There are no alignment requirements.
 */
static const void *body(MD5_CTX *ctx, const void *data, unsigned long size)
{
	const unsigned char *ptr;
	MD5_u32plus a, b, c, d;
	MD5_u32plus saved_a, saved_b, saved_c, saved_d;

	ptr = (const unsigned char *)data;

	a = ctx->a;
	b = ctx->b;
	c = ctx->c;
	d = ctx->d;

	do {
		saved_a = a;
		saved_b = b;
		saved_c = c;
		saved_d = d;

/* Round 1 */
		STEP(F, a, b, c, d, SET(0), 0xd76aa478, 7)
		STEP(F, d, a, b, c, SET(1), 0xe8c7b756, 12)
		STEP(F, c, d, a, b, SET(2), 0x242070db, 17)
		STEP(F, b, c, d, a, SET(3), 0xc1bdceee, 22)
		STEP(F, a, b, c, d, SET(4), 0xf57c0faf, 7)
		STEP(F, d, a, b, c, SET(5), 0x4787c62a, 12)
		STEP(F, c, d, a, b, SET(6), 0xa8304613, 17)
		STEP(F, b, c, d, a, SET(7), 0xfd469501, 22)
		STEP(F, a, b, c, d, SET(8), 0x698098d8, 7)
		STEP(F, d, a, b, c, SET(9), 0x8b44f7af, 12)
		STEP(F, c, d, a, b, SET(10), 0xffff5bb1, 17)
		STEP(F, b, c, d, a, SET(11), 0x895cd7be, 22)
		STEP(F, a, b, c, d, SET(12), 0x6b901122, 7)
		STEP(F, d, a, b, c, SET(13), 0xfd987193, 12)
		STEP(F, c, d, a, b, SET(14), 0xa679438e, 17)
		STEP(F, b, c, d, a, SET(15), 0x49b40821, 22)

/* Round 2 */
		STEP(G, a, b, c, d, GET(1), 0xf61e2562, 5)
		STEP(G, d, a, b, c, GET(6), 0xc040b340, 9)
		STEP(G, c, d, a, b, GET(11), 0x265e5a51, 14)
		STEP(G, b, c, d, a, GET(0), 0xe9b6c7aa, 20)
		STEP(G, a, b, c, d, GET(5), 0xd62f105d, 5)
		STEP(G, d, a, b, c, GET(10), 0x02441453, 9)
		STEP(G, c, d, a, b, GET(15), 0xd8a1e681, 14)
		STEP(G, b, c, d, a, GET(4), 0xe7d3fbc8, 20)
		STEP(G, a, b, c, d, GET(9), 0x21e1cde6, 5)
		STEP(G, d, a, b, c, GET(14), 0xc33707d6, 9)
		STEP(G, c, d, a, b, GET(3), 0xf4d50d87, 14)
		STEP(G, b, c, d, a, GET(8), 0x455a14ed, 20)
		STEP(G, a, b, c, d, GET(13), 0xa9e3e905, 5)
		STEP(G, d, a, b, c, GET(2), 0xfcefa3f8, 9)
		STEP(G, c, d, a, b, GET(7), 0x676f02d9, 14)
		STEP(G, b, c, d, a, GET(12), 0x8d2a4c8a, 20)

/* Round 3 */
		STEP(H, a, b, c, d, GET(5), 0xfffa3942, 4)
		STEP(H2, d, a, b, c, GET(8), 0x8771f681, 11)
		STEP(H, c, d, a, b, GET(11), 0x6d9d6122, 16)
		STEP(H2, b, c, d, a, GET(14), 0xfde5380c, 23)
		STEP(H, a, b, c, d, GET(1), 0xa4beea44, 4)
		STEP(H2, d, a, b, c, GET(4), 0x4bdecfa9, 11)
		STEP(H, c, d, a, b, GET(7), 0xf6bb4b60, 16)
		STEP(H2, b, c, d, a, GET(10), 0xbebfbc70, 23)
		STEP(H, a, b, c, d, GET(13), 0x289b7ec6, 4)
		STEP(H2, d, a, b, c, GET(0), 0xeaa127fa, 11)
		STEP(H, c, d, a, b, GET(3), 0xd4ef3085, 16)
		STEP(H2, b, c, d, a, GET(6), 0x04881d05, 23)
		STEP(H, a, b, c, d, GET(9), 0xd9d4d039, 4)
		STEP(H2, d, a, b, c, GET(12), 0xe6db99e5, 11)
		STEP(H, c, d, a, b, GET(15), 0x1fa27cf8, 16)
		STEP(H2, b, c, d, a, GET(2), 0xc4ac5665, 23)

/* Round 4 */
		STEP(I, a, b, c, d, GET(0), 0xf4292244, 6)
		STEP(I, d, a, b, c, GET(7), 0x432aff97, 10)
		STEP(I, c, d, a, b, GET(14), 0xab9423a7, 15)
		STEP(I, b, c, d, a, GET(5), 0xfc93a039, 21)
		STEP(I, a, b, c, d, GET(12), 0x655b59c3, 6)
		STEP(I, d, a, b, c, GET(3), 0x8f0ccc92, 10)
		STEP(I, c, d, a, b, GET(10), 0xffeff47d, 15)
		STEP(I, b, c, d, a, GET(1), 0x85845dd1, 21)
		STEP(I, a, b, c, d, GET(8), 0x6fa87e4f, 6)
		STEP(I, d, a, b, c, GET(15), 0xfe2ce6e0, 10)
		STEP(I, c, d, a, b, GET(6), 0xa3014314, 15)
		STEP(I, b, c, d, a, GET(13), 0x4e0811a1, 21)
		STEP(I, a, b, c, d, GET(4), 0xf7537e82, 6)
		STEP(I, d, a, b, c, GET(11), 0xbd3af235, 10)
		STEP(I, c, d, a, b, GET(2), 0x2ad7d2bb, 15)
		STEP(I, b, c, d, a, GET(9), 0xeb86d391, 21)

		a += saved_a;
		b += saved_b;
		c += saved_c;
		d += saved_d;

		ptr += 64;
	} while (size -= 64);

	ctx->a = a;
	ctx->b = b;
	ctx->c = c;
	ctx->d = d;

	return ptr;
}

void MD5_Init(MD5_CTX *ctx)
{
	ctx->a = 0x67452301;
	ctx->b = 0xefcdab89;
	ctx->c = 0x98badcfe;
	ctx->d = 0x10325476;

	ctx->lo = 0;
	ctx->hi = 0;
}

void MD5_Update(MD5_CTX *ctx, const void *data, unsigned long size)
{
	MD5_u32plus saved_lo;
	unsigned long used, available;

	saved_lo = ctx->lo;
	if ((ctx->lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
		ctx->hi++;
	ctx->hi += size >> 29;

	used = saved_lo & 0x3f;

	if (used) {
		available = 64 - used;

		if (size < available) {
			memcpy(&ctx->buffer[used], data, size);
			return;
		}

		memcpy(&ctx->buffer[used], data, available);
		data = (const unsigned char *)data + available;
		size -= available;
		body(ctx, ctx->buffer, 64);
	}

	if (size >= 64) {
		data = body(ctx, data, size & ~(unsigned long)0x3f);
		size &= 0x3f;
	}

	memcpy(ctx->buffer, data, size);
}

#define OUT(dst, src) \
	(dst)[0] = (unsigned char)(src); \
	(dst)[1] = (unsigned char)((src) >> 8); \
	(dst)[2] = (unsigned char)((src) >> 16); \
	(dst)[3] = (unsigned char)((src) >> 24);

void MD5_Final(unsigned char *result, MD5_CTX *ctx)
{
	unsigned long used, available;

	used = ctx->lo & 0x3f;

	ctx->buffer[used++] = 0x80;

	available = 64 - used;

	if (available < 8) {
		memset(&ctx->buffer[used], 0, available);
		body(ctx, ctx->buffer, 64);
		used = 0;
		available = 64;
	}

	memset(&ctx->buffer[used], 0, available - 8);

	ctx->lo <<= 3;
	OUT(&ctx->buffer[56], ctx->lo)
	OUT(&ctx->buffer[60], ctx->hi)

	body(ctx, ctx->buffer, 64);

	OUT(&result[0], ctx->a)
	OUT(&result[4], ctx->b)
	OUT(&result[8], ctx->c)
	OUT(&result[12], ctx->d)

	memset(ctx, 0, sizeof(*ctx));
}

#endif
